README
XRoar - a Dragon and Tandy 8-bit computer emulator
Copyright 2003-2023 Ciaran Anscomb <xroar@6809.org.uk>


Introduction
============

XRoar emulates the Dragon 32/64; Tandy Colour Computers 1, 2 and 3; the
Tandy MC-10; and some other similar machines or clones.  It runs on a
wide variety of platforms.  Emulated hardware includes:

   * Dragon 32, 64, and 200-E; Tandy CoCo 1, 2, & 3; Tandy MC-10; Matra
     & Hachette Alice 4K.
   * DragonDOS, Delta and RS-DOS disk controller cartridges.
   * Orchestra 90-CC stereo sound cartridge.
   * Games Master Cartridge, including the SN76489 sound chip.
   * Glenside IDE cartridge, with IDE hard disk image support.
   * NX32 and MOOH RAM expansions, with SPI and SD card image support.

   Other features include:

   * Raw and translated keyboard modes.
   * Read and write tape images (compact '.cas' files or audio, e.g.
     '.wav').
   * Read and write VDK, JVC and DMK format floppy disk images.
   * Becker port for communication with remote servers.
   * Save and load machine snapshots.
   * GDB target for remote debugging.

   XRoar is easily built from source under Linux, and binary packages
are provided for Mac OS X+ and Windows.

   XRoar can also be compiled to WebAssembly, and redistributing it in
this form may provide a convenient way for users to run your Dragon
software.  See XRoar Online (https://www.6809.org.uk/xroar/online/) for
an example.

   XRoar is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

   XRoar is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

   You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

   This README contains extracts from the manual.  Binary packages
contain the full manual as a PDF, which is also available from the XRoar
home page (https://www.6809.org.uk/xroar/).

Recent changes
==============

Version 1.4 replaces the '-ccr simulated' cross-colour renderer with
more CPU-intensive code that also handles PAL. The old NTSC-only
renderer is still available using '-ccr partial'.  Some video options
can be changed on the fly in a new TV Controls dialog.

   Version 1.3 changes the default floppy disk write-back behaviour.
The old behaviour erred on the side of protecting image files from
accidental modification.  Enough people have complained about this--or
at least, the small number that have complained have done so
loudly--that XRoar will now rewrite changes to the backing file by
default.  Run with the '-no-disk-write-back' option to revert to the old
behaviour.  XRoar will still rename the old version of a file to have a
'.bak' extension if possible, and also tries harder not to rewrite the
file if no writes have occurred.

   Version 1.0 introduced support for the Tandy Colour Computer 3 and
the Tandy MC-10.  Version 1.1 adds proper support for the MC-10's French
cousin, the Matra & Hachette Alice (4K).

   Snapshots now store much more state, and of course support the new
emulated machines, but this means the format had to change.  Snapshots
from the last 0.x release are still recognised, and can be loaded, but
this support is likely to be removed in time.

   Tape emulation now supports manual pause control.  On the MC-10, this
defaults to paused, as it has no ability to remotely control the tape
motor.  You will need to un-pause after typing 'CLOAD' or 'CLOADM' on
the MC-10 (File -> Cassette -> Play, or from the tape control tool;
autorunning will do this automatically).

   Previously, the Glenside IDE controller would use a fixed HD image
file in the current working directory.  You must now specify an image
with the '-load-hd0' option.  You can also now attach a second hard disk
image with '-load-hd1', if you have software that can access it.

   Similarly, the NX32 and MOOH cartridges would use a fixed SD image
file, and you must now specify it with the '-load-hd0' option.

   Old IDE images, including those created by XRoar, will have a '.img'
file extension.  In order to distinguish these files from similar images
with no header information, you should now rename these to have a '.ide'
extension.


Installation
************

Binary packages
===============

Pre-built binary packages are available from the XRoar home page
(https://www.6809.org.uk/xroar/).  If one is not available for your
architecture, you will need to build from source.  XRoar should build
and run on any POSIX-like system for which SDL version 2 is available.

   You will also need BASIC ROM images--binary dumps of the firmware
from an original machine.  The originals were part-written by Microsoft,
so they are not distributed in the XRoar packages.

Mac OS X+ binary package
------------------------

Download and unzip the appropriate '.zip' distribution for your system.
Drag the application icon to '/Applications/'.

   For troubleshooting or testing options, it's often a good idea to run
from the command line, but application packages don't make that trivial.
A symbolic link to somewhere in your 'PATH' is all that's required.
e.g.:

     $ sudo ln -s /Applications/XRoar.app/Contents/MacOS/xroar \
             /usr/local/bin/xroar

   After this, you can start the emulator by simply typing 'xroar'
followed by any command line options.

   ROM images should be placed in a directory you create under your
'HOME' named '~/Library/XRoar/roms/' (not the system directory,
'/Library/').  Name any configuration file you create
'~/Library/XRoar/xroar.conf'.

   The Mac OS X+ build provides a menu for access to certain features,
and often accepts the more familiar '<Command>+KEY' in place of the
'<CTRL>+KEY' shortcuts listed in this manual.

Windows binary package
----------------------

Download and unzip the appropriate '.zip' distribution for your system.

   The easiest way forward is to simply put ROM images into the
directory created when you unzip the distribution, and then run the
'.exe' straight from there.  You can also put any configuration file
('xroar.conf') here.

   However, if you want to avoid having to move files around each time
you upgrade, you can create 'Documents/XRoar' to contain your
configuration file, and a subdirectory of that, 'Documents/XRoar/roms'
for ROM images.

   Note when troubleshooting that the logging from the Windows binary is

ChangeLog
Version 1.4.2, Tue 16 Jun 2023

 * Fix running with -ui null

Version 1.4.1, Tue  6 Jun 2023

 * Fix composite colour switching in -ccr simple and -ccr 5bit [Keith Palmer]

Version 1.4, Mon  5 Jun 2023

 * NTSC colour tweaks for slightly better "cyan"
 * Initial support for middle-button paste
 * Add brightness, contrast, saturation & hue controls to GTK+ and Windows UIs
 * Add gain/volume controls to GTK+ and Windows UIs
 * New options: -vo-brightness, -vo-contrast, -vo-colour, -vo-hue
 * New tape control & drive control windows in Windows UI
 * GIME: implement bug in 16-colour, 16 byte-per-row modes
 * New option -vo-pixel-fmt specifies pixel layout and bit depth
 * Old NTSC-only composite renderer renamed -ccr partial
 * New composite renderer that handles PAL used for -ccr simulated
 * WD279x: write sync bytes before DAM (fixes written-to DMK) [Michael Furman]
 * New option: -no-vo-colour-killer

Version 1.3.1, Tue 14 Feb 2023

 * Further GIME VSC tweaks
 * Allow preemption of Control key (for CoCo 3)
 * Fix GTK+ zoom in/out
 * Work around build issues on Apple Silicon [Allen Huffman, Blair Leduc]

Version 1.3, Wed  4 Jan 2023

 * Add -no-ratelimit option to start at maximum speed
 * Further fixes to 6809 TFR/EXG involving CC/DP [Tim Lindner]
 * Fix Delta density select [Phill Harvey-Smith]
 * Track floppy disk 'dirty' state to avoid unnecessary image rewrites
 * Important: disk write-back now defaults to ENABLED
 * Better Vertical SCroll register behaviour in GIME [Ralph Serpas]
 * 6309 timing fix for bit operations

Version 1.2, Thu 27 Oct 2022

 * Fixed comma, lowercase 'm', lowercase 'ø' glyphs for GIME [Tim Lindner]
 * Fix SDL-only builds
 * 6809: flesh out some illegal instruction behaviours [David Banks]
 * 6309: flesh out some undocumented behaviour [David Banks]
 * Fleshed out T1-compatibility in CoCo 3 GIME [R. Allen Murphy]

Version 1.1, Thu 21 Jul 2022

 * New GDB monitor commands [by Tormod Volden]
 * New configure options to only build specific machine archs
 * Support 1M or 2M in CoCo 3 [with Christopher Hawks]
 * Support K7 cassette image files (read-only)
 * Support UTF-8 block characters in -type for MC-10
 * Type ASCII BASIC from file on MC-10
 * Matra & Hachette Alice support (keyboard layout, built-in profile)
 * New meta-options -machine-opt and -cart-opt
 * New ide-addr=address cart-opt
 * New abstract block device handling
 * IDE support adjusted to use abstracted block devices
 * MOOH/NX32 support adjusted to use abstracted block devices
 * 6801/6803: fix some illegal instruction timings [George Phillips]
 * Fixed uppercase 'G', lowercase 'j' and 'w' glyphs for 6847T1 [Tim Lindner]

Version 1.0.9, Tue 18 Jan 2022

 * Close file after serialisation [Steve Rasmussen]

Version 1.0.8, Thu 13 Jan 2022

 * Fix WASM audio for non-Firefox [Greg Dionne]
 * Fix crash reading zero-length CAS file
 * Fix joystick reads for Pacdude Monster Maze

Version 1.0.7, Sat  1 Jan 2022

 * Fix Windows 11 video (set SDL hint to use different renderer)

Version 1.0.6, Fri 31 Dec 2021

 * Fix SDL audio thread interactions
 * Try multiple SDL video renderers in order
 * Fix composite phase setting

Version 1.0.5, Thu 23 Dec 2021

 * Open IDE images in binary mode under Windows [by Tormod Volden]

Version 1.0.4, Sun 12 Dec 2021


machine.c
/*  Copyright 2003-2013 Ciaran Anscomb
 *
 *  This file is part of XRoar.
 *
 *  XRoar is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  XRoar is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XRoar.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <assert.h>
#include <fcntl.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "portalib/glib.h"
#include "portalib/strings.h"

#include "cart.h"
#include "crc32.h"
#include "fs.h"
#include "hd6309.h"
#include "hd6309_trace.h"
#include "input.h"
#include "joystick.h"
#include "keyboard.h"
#include "logging.h"
#include "machine.h"
#include "mc6809.h"
#include "mc6809_trace.h"
#include "mc6821.h"
#include "module.h"
#include "path.h"
#include "printer.h"
#include "romlist.h"
#include "sam.h"
#include "sound.h"
#include "tape.h"
#include "vdg.h"
#include "vdrive.h"
#include "wd279x.h"
#include "xroar.h"

unsigned int machine_ram_size = 0x10000;  /* RAM in bytes, up to 64K */
uint8_t machine_ram[0x10000];
static uint8_t *machine_rom;
static uint8_t rom0[0x4000];
static uint8_t rom1[0x4000];
struct MC6809 *CPU0 = NULL;
struct MC6821 PIA0, PIA1;
struct cart *machine_cart = NULL;
static struct cart running_cart;
_Bool has_bas, has_extbas, has_altbas;
uint32_t crc_bas, crc_extbas, crc_altbas;

/* Useful configuration side-effect tracking */
static _Bool unexpanded_dragon32 = 0;
static enum {
	RAM_ORGANISATION_4K,
	RAM_ORGANISATION_16K,
	RAM_ORGANISATION_64K
} ram_organisation = RAM_ORGANISATION_64K;
static uint16_t ram_mask = 0xffff;
static _Bool have_acia = 0;

static struct {
	const char *bas;
	const char *extbas;
	const char *altbas;
} rom_list[] = {
	{ NULL, "@dragon32", NULL },
	{ NULL, "@dragon64", "@dragon64_alt" },
	{ "@coco", "@coco_ext", NULL }
};

struct machine_config_template {
	const char *name;
	const char *description;
	int architecture;
	int tv_standard;
	int ram;
};

static struct machine_config_template config_templates[] = {
	{ "dragon32", "Dragon 32", ARCH_DRAGON32, TV_PAL, 32, },
	{ "dragon64", "Dragon 64", ARCH_DRAGON64, TV_PAL, 64, },
	{ "tano", "Tano Dragon (NTSC)", ARCH_DRAGON64, TV_NTSC, 64, },
	{ "coco", "Tandy CoCo (PAL)", ARCH_COCO, TV_PAL, 64, },
	{ "cocous", "Tandy CoCo (NTSC)", ARCH_COCO, TV_NTSC, 64, },
};
#define NUM_CONFIG_TEMPLATES (int)(sizeof(config_templates)/sizeof(struct machine_config_template))

static struct machine_config **configs = NULL;
static int num_configs = NUM_CONFIG_TEMPLATES;

static void initialise_ram(void);

static int cycles;
static uint8_t read_cycle(uint16_t A);
static void write_cycle(uint16_t A, uint8_t D);
static void vdg_fetch_handler(int nbytes, uint8_t *dest);

/**************************************************************************/

/* Create config array */
static int alloc_config_array(int size) {
	struct machine_config **new_list;
	int clear_from = num_configs;
	if (!configs) clear_from = 0;
	new_list = g_realloc(configs, size * sizeof(struct machine_config *));
	configs = new_list;
	memset(&configs[clear_from], 0, (size - clear_from) * sizeof(struct machine_config *));
	return 0;
}

/* Populate config from template */
static int populate_config_index(int i) {
	assert(configs != NULL);
	assert(i >= 0 && i < NUM_CONFIG_TEMPLATES);
	if (configs[i])
		return 0;
	configs[i] = g_malloc0(sizeof(struct machine_config));
	configs[i]->name = g_strdup(config_templates[i].name);
	configs[i]->description = g_strdup(config_templates[i].description);
	configs[i]->architecture = config_templates[i].architecture;
	configs[i]->cpu = CPU_MC6809;
	configs[i]->keymap = ANY_AUTO;
	configs[i]->tv_standard = config_templates[i].tv_standard;
	configs[i]->ram = config_templates[i].ram;
	configs[i]->index = i;
	return 0;
}

struct machine_config *machine_config_new(void) {
	struct machine_config *new;
	if (alloc_config_array(num_configs+1) != 0)
		return NULL;
	new = g_malloc0(sizeof(struct machine_config));
	new->index = num_configs;
	new->architecture = ANY_AUTO;
	new->cpu = CPU_MC6809;
	new->keymap = ANY_AUTO;
	new->tv_standard = ANY_AUTO;
	new->ram = ANY_AUTO;
	configs[num_configs++] = new;
	return new;
}

int machine_config_count(void) {
	return num_configs;
}

struct machine_config *machine_config_index(int i) {
	if (i < 0 || i >= num_configs) {
		return NULL;
	}
	if (!configs) {
		if (alloc_config_array(num_configs) != 0)
			return NULL;
	}
	if (i < NUM_CONFIG_TEMPLATES && !configs[i]) {
		if (populate_config_index(i) != 0)
			return NULL;
	}
	return configs[i];
}

struct machine_config *machine_config_by_name(const char *name) {
	int count, i;
	if (!name) return NULL;
	count = machine_config_count();
	for (i = 0; i < count; i++) {
		struct machine_config *mc = machine_config_index(i);
		if (0 == strcmp(mc->name, name)) {
			return mc;
		}
	}
	return NULL;
}

struct machine_config *machine_config_by_arch(int arch) {
	int count, i;
	count = machine_config_count();
	for (i = 0; i < count; i++) {
		struct machine_config *mc = machine_config_index(i);
		if (mc->architecture == arch) {
			return mc;
		}
	}
	return NULL;
}

static int find_working_arch(void) {
	int arch;
	char *tmp = NULL;
	if ((tmp = romlist_find("@dragon64"))) {
		arch = ARCH_DRAGON64;
	} else if ((tmp = romlist_find("@dragon32"))) {
		arch = ARCH_DRAGON32;
	} else if ((tmp = romlist_find("@coco"))) {
		arch = ARCH_COCO;
	} else {
		/* Fall back to this, which won't start up properly */
		arch = ARCH_DRAGON64;
	}
	if (tmp)
		g_free(tmp);
	return arch;
}

struct machine_config *machine_config_first_working(void) {
	return machine_config_by_arch(find_working_arch());
}

void machine_config_complete(struct machine_config *mc) {
	if (!mc->description) {
		mc->description = g_strdup(mc->name);
	}
	if (mc->tv_standard == ANY_AUTO)
		mc->tv_standard = TV_PAL;
	/* Various heuristics to find a working architecture */
	if (mc->architecture == ANY_AUTO) {
		/* TODO: checksum ROMs to help determine arch */
		if (mc->bas_rom) {
			mc->architecture = ARCH_COCO;
		} else if (mc->altbas_rom) {
			mc->architecture = ARCH_DRAGON64;
		} else if (mc->extbas_rom) {
			struct stat statbuf;
			mc->architecture = ARCH_DRAGON64;
			if (stat(mc->extbas_rom, &statbuf) == 0) {
				if (statbuf.st_size <= 0x2000) {
					mc->architecture = ARCH_COCO;
				}
			}
		} else {
			mc->architecture = find_working_arch();
		}
	}
	if (mc->ram < 4 || mc->ram > 64) {
		switch (mc->architecture) {
			case ARCH_DRAGON32:
				mc->ram = 32;
				break;
			default:
				mc->ram = 64;
				break;
		}
	}
	if (mc->keymap == ANY_AUTO) {
		switch (mc->architecture) {
		case ARCH_DRAGON64: case ARCH_DRAGON32: default:
			mc->keymap = KEYMAP_DRAGON;
			break;
		case ARCH_COCO:
			mc->keymap = KEYMAP_COCO;
			break;
		}
	}
	/* Now find which ROMs we're actually going to use */
	if (!mc->nobas && !mc->bas_rom && rom_list[mc->architecture].bas) {
		mc->bas_rom = g_strdup(rom_list[mc->architecture].bas);
	}
	if (!mc->noextbas && !mc->extbas_rom && rom_list[mc->architecture].extbas) {
		mc->extbas_rom = g_strdup(rom_list[mc->architecture].extbas);
	}
	if (!mc->noaltbas && !mc->altbas_rom && rom_list[mc->architecture].altbas) {
		mc->altbas_rom = g_strdup(rom_list[mc->architecture].altbas);
	}
}

/* ---------------------------------------------------------------------- */

static void pia0a_data_preread(void) {
	keyboard_update();
	PIA0.a.in_sink &= input_firebutton_mask;
}

#define pia0a_data_postwrite NULL
#define pia0a_control_postwrite sound_update

#define pia0b_data_preread keyboard_update
#define pia0b_data_postwrite NULL
#define pia0b_control_postwrite sound_update

static void pia0b_data_preread_coco64k(void) {
	keyboard_update();
	/* PB6 of PIA0 is linked to PB2 of PIA1 on 64K CoCos */
	if (PIA_VALUE_B(PIA1) & (1<<2)) {
		PIA0.b.in_source |= (1<<6);
		PIA0.b.in_sink |= (1<<6);
	} else {
		PIA0.b.in_source &= ~(1<<6);
		PIA0.b.in_sink &= ~(1<<6);
	}
}

static void pia1b_data_preread_coco64k(void) {
	keyboard_update();
	/* PB6 of PIA0 is linked to PB2 of PIA1 on 64K CoCos */
	if (PIA_VALUE_B(PIA0) & (1<<6)) {
		PIA1.b.in_source |= (1<<2);
		PIA1.b.in_sink |= (1<<2);
	} else {
		PIA1.b.in_source &= ~(1<<2);
		PIA1.b.in_sink &= ~(1<<2);
	}
}

#define pia1a_data_preread NULL

static void pia1a_data_postwrite(void) {
	sound_update();
	joystick_update();
	tape_update_output();
	if (IS_DRAGON) {
		keyboard_update();
		printer_strobe();
	}
}

#define pia1a_control_postwrite tape_update_motor

#define pia1b_data_preread NULL

static void pia1b_data_postwrite(void) {
	if (IS_DRAGON64) {
		machine_rom = PIA_VALUE_B(PIA1) & 0x04 ? rom0 : rom1;
	}
	sound_update();
	vdg_set_mode(PIA1.b.out_source & PIA1.b.out_sink);
}
#define pia1b_control_postwrite sound_update

void machine_init(void) {
	sam_init();
	mc6821_init(&PIA0);
	PIA0.a.data_preread = pia0a_data_preread;
	PIA0.a.data_postwrite = pia0a_data_postwrite;
	PIA0.b.data_preread = pia0b_data_preread;
	PIA0.b.data_postwrite = pia0b_data_postwrite;
	mc6821_init(&PIA1);
	PIA1.a.data_preread = pia1a_data_preread;
	PIA1.a.data_postwrite = pia1a_data_postwrite;
	PIA1.a.control_postwrite = pia1a_control_postwrite;
	PIA1.b.data_preread = pia1b_data_preread;
	PIA1.b.data_postwrite = pia1b_data_postwrite;
#ifndef FAST_SOUND
	machine_select_fast_sound(xroar_fast_sound);
#endif
	vdrive_init();
	vdg_init();
	tape_init();

	vdg_fetch_bytes = vdg_fetch_handler;
}

void machine_shutdown(void) {
	tape_shutdown();
	vdrive_shutdown();
}

void machine_configure(struct machine_config *mc) {
	if (!mc) return;
	machine_config_complete(mc);
	if (mc->description) {
		LOG_DEBUG(2, "Machine: %s\n", mc->description);
	}
	/* */
	if (CPU0) {
		CPU0->free(CPU0);
		CPU0 = NULL;
	}
	switch (mc->cpu) {
	case CPU_MC6809: default:
		CPU0 = mc6809_new();
		break;
	case CPU_HD6309:
		CPU0 = hd6309_new();
		break;
	}
	CPU0->read_cycle = read_cycle;
	CPU0->write_cycle = write_cycle;
	xroar_set_keymap(mc->keymap);
	switch (mc->tv_standard) {
	case TV_PAL: default:
		xroar_select_cross_colour(CROSS_COLOUR_OFF);
		break;
	case TV_NTSC:
		xroar_select_cross_colour(CROSS_COLOUR_KBRW);
		break;
	}
	/* Load appropriate ROMs */
	memset(rom0, 0, sizeof(rom0));
	memset(rom1, 0, sizeof(rom1));
	/* ... BASIC */
	has_bas = 0;
	crc_bas = 0;
	if (!mc->nobas && mc->bas_rom) {
		char *tmp = romlist_find(mc->bas_rom);
		if (tmp) {
			int size = machine_load_rom(tmp, rom0 + 0x2000, sizeof(rom0) - 0x2000);
			if (size > 0) {
				has_bas = 1;
				crc_bas = crc32_block(CRC32_RESET, rom0 + 0x2000, size);
				LOG_DEBUG(2, "\tCRC = 0x%08x\n", crc_bas);
				if (IS_COCO && xroar_opt_force_crc_match) {
					/* force Color BASIC 1.3 */
					crc_bas = 0xd8f4d15e;
				}
			}
			g_free(tmp);
		}
	}
	/* ... Extended BASIC */
	has_extbas = 0;
	crc_extbas = 0;
	if (!mc->noextbas && mc->extbas_rom) {
		char *tmp = romlist_find(mc->extbas_rom);
		if (tmp) {
			int size = machine_load_rom(tmp, rom0, sizeof(rom0));
			if (size > 0) {
				has_extbas = 1;
				crc_extbas = crc32_block(CRC32_RESET, rom0, size);
				LOG_DEBUG(2, "\tCRC = 0x%08x\n", crc_extbas);
				if (xroar_opt_force_crc_match) {
					if (IS_DRAGON64) {
						/* force Dragon 64 BASIC */
						crc_extbas = 0x84f68bf9;
					} else if (IS_DRAGON32) {
						/* force Dragon 32 BASIC */
						crc_extbas = 0xe3879310;
					}
				}
			}
			if (size > 0x2000) {
				has_bas = 1;
				crc_bas = crc32_block(CRC32_RESET, rom0 + 0x2000, size - 0x2000);
			}
			g_free(tmp);
		}
	}
	/* ... Alternate BASIC ROM */
	has_altbas = 0;
	crc_altbas = 0;
	if (!mc->noaltbas && mc->altbas_rom) {
		char *tmp = romlist_find(mc->altbas_rom);
		if (tmp) {
			int size = machine_load_rom(tmp, rom1, sizeof(rom1));
			if (size > 0) {
				has_altbas = 1;
				crc_altbas = crc32_block(CRC32_RESET, rom1, size);
				LOG_DEBUG(2, "\tCRC = 0x%08x\n", crc_altbas);
			}
			g_free(tmp);
		}
	}
	machine_ram_size = mc->ram * 1024;
	/* This will be under PIA control on a Dragon 64 */
	machine_rom = rom0;

	/* Default all PIA connections to unconnected (no source, no sink) */
	PIA0.b.in_source = 0;
	PIA1.b.in_source = 0;
	PIA0.a.in_sink = PIA0.b.in_sink = 0xff;
	PIA1.a.in_sink = PIA1.b.in_sink = 0xff;
	/* Machine-specific PIA connections */
	if (IS_DRAGON) {
		/* Centronics printer port - !BUSY */
		PIA1.b.in_source |= (1<<0);
	}
	if (IS_DRAGON64) {
		have_acia = 1;
		PIA1.b.in_source |= (1<<2);
	} else if (IS_COCO && machine_ram_size <= 0x1000) {
		/* 4K CoCo ties PB2 of PIA1 low */
		PIA1.b.in_sink &= ~(1<<2);
	} else if (IS_COCO && machine_ram_size <= 0x4000) {
		/* 16K CoCo pulls PB2 of PIA1 high */
		PIA1.b.in_source |= (1<<2);
	}
	PIA0.b.data_postwrite = pia0b_data_postwrite;
	PIA0.b.data_preread = pia0b_data_preread;
	PIA1.b.data_preread = pia1b_data_preread;
	if (IS_COCO && machine_ram_size > 0x4000) {
		/* 64K CoCo connects PB6 of PIA0 to PB2 of PIA1.
		 * Deal with this through a postwrite. */
		PIA0.b.data_preread = pia0b_data_preread_coco64k;
		PIA1.b.data_preread = pia1b_data_preread_coco64k;
	}

	unexpanded_dragon32 = 0;
	ram_mask = 0xffff;

	if (IS_COCO) {
		if (machine_ram_size <= 0x2000) {
			ram_organisation = RAM_ORGANISATION_4K;
			ram_mask = 0x3f3f;
		} else if (machine_ram_size <= 0x4000) {
			ram_organisation = RAM_ORGANISATION_16K;
		} else {
			ram_organisation = RAM_ORGANISATION_64K;
			if (machine_ram_size <= 0x8000)
				ram_mask = 0x7fff;
		}
	}

	if (IS_DRAGON) {
		if (IS_DRAGON32 && machine_ram_size <= 0x8000) {
			unexpanded_dragon32 = 1;
			ram_mask = 0x7fff;
		}
	}
}

void machine_reset(_Bool hard) {
	if (hard) {
		initialise_ram();
	}
	mc6821_reset(&PIA0);
	mc6821_reset(&PIA1);
	if (machine_cart && machine_cart->reset) {
		machine_cart->reset();
	}
	sam_reset();
	CPU0->reset(CPU0);
#ifdef TRACE
	mc6809_trace_reset();
	hd6309_trace_reset();
#endif
	vdg_reset();
	tape_reset();
}

void machine_run(int ncycles) {
	cycles += ncycles;
	CPU0->running = 1;
	CPU0->run(CPU0);
}

static uint16_t decode_Z(uint16_t Z) {
	switch (ram_organisation) {
	case RAM_ORGANISATION_4K:
		return (Z & 0x3f) | ((Z & 0x3f00) >> 2) | ((~Z & 0x8000) >> 3);
	case RAM_ORGANISATION_16K:
		return (Z & 0x7f) | ((Z & 0x7f00) >> 1) | ((~Z & 0x8000) >> 1);
	case RAM_ORGANISATION_64K: default:
		return Z & ram_mask;
	}
}

/* Interface to SAM to decode and translate address */
static _Bool do_cpu_cycle(uint16_t A, _Bool RnW, int *S, uint16_t *Z) {
	uint16_t tmp_Z;
	int ncycles;
	_Bool is_ram_access = sam_run(A, RnW, S, &tmp_Z, &ncycles);
	if (is_ram_access) {
		*Z = decode_Z(tmp_Z);
	}
	cycles -= ncycles;
	if (cycles <= 0) CPU0->running = 0;
	event_current_tick += ncycles;
	event_run_queue(MACHINE_EVENT_LIST);
	MC6809_IRQ_SET(CPU0, PIA0.a.irq | PIA0.b.irq);
	MC6809_FIRQ_SET(CPU0, PIA1.a.irq | PIA1.b.irq);
	return is_ram_access;
}

static uint8_t read_D = 0;

static uint8_t read_cycle(uint16_t A) {
	int S;
	uint16_t Z;
	_Bool is_ram_access = do_cpu_cycle(A, 1, &S, &Z);
	/* Thanks to CrAlt on #coco_chat for verifying that RAM accesses
	 * produce a different "null" result on his 16K CoCo */
	if (is_ram_access)
		read_D = 0xff;
	switch (S) {
		case 0:
			if (Z < machine_ram_size)
				read_D = machine_ram[Z];
			break;
		case 1:
		case 2:
			read_D = machine_rom[A & 0x3fff];
			break;
		case 3:
			if (machine_cart)
				read_D = machine_cart->mem_data[A & 0x3fff];
			break;
		case 4:
			if (IS_COCO) {
				read_D = mc6821_read(&PIA0, A & 3);
			} else {
				if ((A & 4) == 0) {
					read_D = mc6821_read(&PIA0, A & 3);
				} else {
					if (have_acia) {
						/* XXX Dummy ACIA reads */
						switch (A & 3) {
						default:
						case 0:  /* Receive Data */
						case 3:  /* Control */
							read_D = 0x00;
							break;
						case 2:  /* Command */
							read_D = 0x02;
							break;
						case 1:  /* Status */
							read_D = 0x10;
							break;
						}
					}
				}
			}
			break;
		case 5:
			read_D = mc6821_read(&PIA1, A & 3);
			break;
		case 6:
			if (machine_cart && machine_cart->io_read) {
				read_D = machine_cart->io_read(A);
			}
			break;
		default:
			break;
	}
#ifdef TRACE
	if (xroar_trace_enabled) {
		switch (xroar_machine_config->cpu) {
		case CPU_MC6809: default:
			mc6809_trace_byte(read_D, A);
			break;
		case CPU_HD6309:
			hd6309_trace_byte(read_D, A);
			break;
		}
	}
#endif
	return read_D;
}

static void write_cycle(uint16_t A, uint8_t D) {
	int S;
	uint16_t Z;
	_Bool is_ram_access = do_cpu_cycle(A, 0, &S, &Z);
	if ((S & 4) || unexpanded_dragon32) {
		switch (S) {
			case 1:
			case 2:
				D = machine_rom[A & 0x3fff];
				break;
			case 3:
				if (machine_cart)
					D = machine_cart->mem_data[A & 0x3fff];
				break;
			case 4:
				if (IS_COCO) {
					mc6821_write(&PIA0, A & 3, D);
				} else {
					if ((A & 4) == 0) {
						mc6821_write(&PIA0, A & 3, D);
					}
				}
				break;
			case 5:
				mc6821_write(&PIA1, A & 3, D);
				break;
			case 6:
				if (machine_cart && machine_cart->io_write) {
					machine_cart->io_write(A, D);
				}
				break;
			default:
				break;
		}
	}
	if (is_ram_access) {
		machine_ram[Z] = D;
	}
}

static void vdg_fetch_handler(int nbytes, uint8_t *dest) {
	while (nbytes > 0) {
		uint16_t V = 0;
		_Bool valid;
		int n = sam_vdg_bytes(nbytes, &V, &valid);
		if (dest) {
			if (valid) {
				V = decode_Z(V);
			}
			memcpy(dest, machine_ram + V, n);
			dest += n;
		}
		nbytes -= n;
	}
}

/* simplified read byte for use by convenience functions - ROM and RAM only */
uint8_t machine_read_byte(uint16_t A) {
	int S, ncycles;
	uint16_t Z;
	_Bool is_ram_access = sam_run(A, 1, &S, &Z, &ncycles);
	if (is_ram_access) {
		Z = decode_Z(Z);
		if (Z < machine_ram_size)
			return machine_ram[Z];
		return 0;
	}
	if (S == 1 || S == 2)
		return machine_rom[A & 0x3fff];
	if (S == 3 && machine_cart)
		return machine_cart->mem_data[A & 0x3fff];
	return 0;
}

/* simplified write byte for use by convenience functions - RAM only */
void machine_write_byte(uint16_t A, uint8_t D) {
	int S, ncycles;
	uint16_t Z;
	int is_ram_access = sam_run(A, 0, &S, &Z, &ncycles);
	if (!is_ram_access)
		return;
	Z = decode_Z(Z);
	if (Z < machine_ram_size)
		machine_ram[Z] = D;
}

/* simulate an RTS without otherwise affecting machine state */
void machine_op_rts(struct MC6809 *cpu) {
	unsigned int new_pc = machine_read_byte(cpu->reg_s) << 8;
	new_pc |= machine_read_byte(cpu->reg_s + 1);
	cpu->reg_s += 2;
	cpu->reg_pc = new_pc;
}

#ifndef FAST_SOUND
void machine_set_fast_sound(_Bool fast) {
	xroar_fast_sound = fast;
	if (fast) {
		PIA0.a.control_postwrite = NULL;
		PIA0.b.control_postwrite = NULL;
		PIA1.b.control_postwrite = NULL;
	} else  {
		PIA0.a.control_postwrite = pia0a_control_postwrite;
		PIA0.b.control_postwrite = pia0b_control_postwrite;
		PIA1.b.control_postwrite = pia1b_control_postwrite;
	}
}

void machine_select_fast_sound(_Bool fast) {
	if (ui_module->fast_sound_changed_cb) {
		ui_module->fast_sound_changed_cb(fast);
	}
	machine_set_fast_sound(fast);
}
#endif

void machine_insert_cart(struct cart_config *cc) {
	machine_remove_cart();
	if (cc) {
		cart_configure(&running_cart, cc);
		machine_cart = &running_cart;
		if (machine_cart->attach) {
			machine_cart->attach();
		}
	}
}

void machine_remove_cart(void) {
	if (machine_cart && machine_cart->detach) {
		machine_cart->detach();
	}
	machine_cart = NULL;
}

/* Intialise RAM contents */
static void initialise_ram(void) {
	int loc = 0, val = 0xff;
	/* Don't know why, but RAM seems to start in this state: */
	while (loc < 0x10000) {
		machine_ram[loc++] = val;
		machine_ram[loc++] = val;
		machine_ram[loc++] = val;
		machine_ram[loc++] = val;
		if ((loc & 0xff) != 0)
			val ^= 0xff;
	}
}

/**************************************************************************/

int machine_load_rom(const char *path, uint8_t *dest, size_t max_size) {
	char *dot;
	FILE *fd;
	int size;

	if (path == NULL)
		return -1;
	dot = strrchr(path, '.');
	if (!(fd = fopen(path, "rb"))) {
		return -1;
	}
	if (dot && strcasecmp(dot, ".dgn") == 0) {
		LOG_DEBUG(2, "Loading DGN: %s\n", path);
		fread(dest, 1, 16, fd);
	} else {
		LOG_DEBUG(2, "Loading ROM: %s\n", path);
	}
	size = fread(dest, 1, max_size, fd);
	fclose(fd);
	return size;
}
